htlcswitch/link.go

```diff
@@ -1176,6 +1188,20 @@ func (l *channelLink) htlcManager() {
 				return
 			}
 
+		case req := <-l.shutdownRequests:
+			// If the channel is clean, we send nil on the err chan
+			// and return to prevent the htlcManager goroutine from
+			// processing any more updates. The full link shutdown
+			// will be triggered by RemoveLink in the peer.
+			if l.channel.IsChannelClean() {
```

Comment:
Hmm perhaps we need to also flip the `EligibleToForward` bit to false? Otherwise, it's possible that after a co-op close is initiated, the switch attempts to forward to this link, causing an HTLC to effectively get "stuck".

Reply:
Makes sense

Reply:
Given that `RemoveLink` is the responsibility of the caller, nice catch

Reply:
It won't get stuck and will get failed back by the mailbox after the "holding timeout".  But in the interest of quick routing, I'll set this bit to false

Reply:
On second thought, this could already happen if `EligibleToForward` is modified to return false:
- `ShutdownIfChannelClean`
- mark as unable to forward
- link returns from main `htlcManager` loop, effectively being stopped but w/o calling `Stop`
- calling function calls `switch.RemoveLink` which calls `link.Stop`

Between steps 1 & 2, the switch would be allowed to forward an HTLC to the link which then gets stuck until the mailbox times it out.  It is not yet known if the shutdown will succeed so unless we switch steps 1 & 2 (pre-emptively mark as unable to forward), we'll still have the problem.

Reply:
Gotcha, so then either way, the mailbox still protects us here due to the cancel after timeout. 

In either case, don't we still need to always ensure that the link is removed from the switch's index? Scanning this file again, doesn't appear that we properly doe this in any of the scenarios that can cause the link to exit. As is, this only happens when the peer itself is torn down. Otherwise, it's possible for this link to die, but still be in the switch's map, and accept new HTLCs. 

In any case, given that this is an existing issue, I think we can handle it in a follow up PR.